// @fluentui/react-tree/lib/utils/ImmutableSet.js
export const emptyImmutableSet = /*#__PURE__*/createImmutableSet();
/**
 * properly creates an ImmutableSet instance from an iterable
 */
export function createImmutableSet(iterable) {
  const internalSet = new Set(iterable);
  return dangerouslyCreateImmutableSet(internalSet);
}
/**
 * Avoid using *dangerouslyCreateImmutableSet*, since this method will expose internally used set, use  createImmutableSet instead,
 * @param internalSet - a set that is used internally to store values.
 */
export function dangerouslyCreateImmutableSet(internalSet) {
  return {
    size: internalSet.size,
    add(value) {
      const nextSet = new Set(internalSet);
      nextSet.add(value);
      return dangerouslyCreateImmutableSet(nextSet);
    },
    clear() {
      return emptyImmutableSet;
    },
    delete(value) {
      const nextSet = new Set(internalSet);
      nextSet.delete(value);
      return dangerouslyCreateImmutableSet(nextSet);
    },
    has(value) {
      return internalSet.has(value);
    },
    [Symbol.iterator]() {
      return internalSet[Symbol.iterator]();
    }
  };
}
